nav2_map_server/src/occ_grid_loader.cpp

```diff
@@ -152,25 +160,67 @@ nav2_util::CallbackReturn OccGridLoader::on_configure(const rclcpp_lifecycle::St
     RCLCPP_ERROR(
       node_->get_logger(), "Failed to load image file %s for reason: %s",
       loadParameters.image_file_name.c_str(), e.what());
+    response->result = nav2_msgs::srv::LoadMap::Response::RESULT_INVALID_MAP_DATA;
+    return false;
+  }
+  return true;
+}
+
+nav2_util::CallbackReturn OccGridLoader::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Configuring");
 
-    throw std::runtime_error("Failed to load map image file.");
+  // initialize Occupancy Grid msg - needed by loadMapFromYaml
+  msg_ = std::make_unique<nav_msgs::msg::OccupancyGrid>();
+
+  if (!loadMapFromYaml(yaml_filename_)) {
+    throw std::runtime_error("Failed to load map yaml file: " + yaml_filename_);
```

Comment:
If we expose another service to change maps, should we throw here still or log error so that its still alive to have a map changed?

Reply:
This is basically just a refactor of this line of code:
https://github.com/ros-planning/navigation2/blob/c2d93596084968376ccb3d6213621f1cf59fe7b5/nav2_map_server/src/occ_grid_loader.cpp#L141

When I refactored the code out of the `on_configure` into the `LoadMapFromYaml` method, I just refactored the throw to the calling method.

I did this so that the exception will be thrown if we're unable to load a map during startup (on_configure) but not if it fails via the service call. If it fails during the service call it returns a fail code in the response.

Reply:
Sorry maybe I wasnt clear, the situation I'm referring to is on `startup -> bogus yaml file -> throws exception server dies`.

I'd like to have `startup -> bogus yaml file -> errors, all the errors -> server stays up` so that someone can use the service to provide a non-bogus file.

Server throwing the exception made sense before the server had another interface to change files because if it fails on startup, its completely useless. Now that we can change map files, if we detect a problem we can request a map to see what's up or request a new map location.

Reply:
See line 136 for one of the places where the response is set to a fail code:
`    response->result = nav2_msgs::srv::LoadMap::Response::RESULT_MAP_DOES_NOT_EXIST;`

Reply:
Can you re-read my clarification? I'm not talking about the service, I'm talking about the line this comment is on in startup. I don't want the server to die on _startup_, in `on_configure` when the yaml is invalid.

Reply:
That was already the behavior, I didn't change that. If you want that to change from how it works today, that should be a separate PR.

Reply:
If I recall correctly, the same behavior also occurs in ROS1.

Reply:
@SteveMacenski - in order to do what you're proposing, allowing the user to pass in an invalid yaml file (or not pass one in at all), it will also require changing map_server.cpp: (starting at line 54)
```
 // Make sure that there's a valid file there and open it up
  std::ifstream fin(yaml_filename.c_str());
  if (fin.fail()) {
    throw std::runtime_error("Could not open '" + yaml_filename + "': file not found");
  }

  // The YAML document from which to get the conversion parameters
  YAML::Node doc = YAML::LoadFile(yaml_filename);

  // Get the map type so that we can create the correct map loader
  std::string map_type;
  try {
    map_type = doc["map_type"].as<std::string>();
  } catch (YAML::Exception &) {
    // Default to occupancy grid if not specified in the YAML file
    map_type = "occupancy";
  }

  // Create the correct map loader for the specified map type
  if (map_type == "occupancy") {
    map_loader_ = std::make_unique<OccGridLoader>(shared_from_this(), yaml_filename);
  } else {
    std::string msg = "Cannot load unknown map type: '" + map_type + "'";
    throw std::runtime_error(msg);
  }
```

I **could** make those additional changes, but I don't know if that will cause other issues and tests to fail. This is why I'm saying that its fine as a future PR, but outside the scope of this one.

Reply:
Note that in the code above it parses the yaml file to determine the map_type of "occupancy" before deciding to create the map_loader_ object of type OccGridLoader. This was done so that we **could** add other map type support in the future. I suppose we could default to "occupancy" if no yaml file is given. Again, I feel this is adding too much to this PR.

Reply:
OK.  I'm not going to scope it into this PR if you don't want to.  

Generally servers dying I look at with the most immediate concern. Especially with service calls and actions if we don't have timeouts (or way too large of timeouts) on the caller. This could block other processes and cause a critical safety failure. I'd rather have a buggy server responding than a dead server

Reply:
Again, the way it is in this PR is NOT CHANGED from the way it was before. IF you try to launch map_server without a valid yaml file as input, it will fail during "on_configure". This is the same as the behavior today in the master branch.